watertap/unit_models/zero_order/tests/test_nanofiltration_zo.py

```diff
@@ -0,0 +1,301 @@
+###############################################################################
+# WaterTAP Copyright (c) 2021, The Regents of the University of California,
+# through Lawrence Berkeley National Laboratory, Oak Ridge National
+# Laboratory, National Renewable Energy Laboratory, and National Energy
+# Technology Laboratory (subject to receipt of any required approvals from
+# the U.S. Dept. of Energy). All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. These files are also available online at the URL
+# "https://github.com/watertap-org/watertap/"
+#
+###############################################################################
+"""
+Tests for zero-order nanofiltration model
+"""
+import pytest
+
+from pyomo.environ import (
+    ConcreteModel, Constraint, SolverStatus, TerminationCondition, value, Var)
+from pyomo.util.check_units import assert_units_consistent
+
+from idaes.core import FlowsheetBlock
+from idaes.core.util import get_solver
+from idaes.core.util.model_statistics import degrees_of_freedom
+from idaes.core.util.testing import initialization_tester
+
+from watertap.unit_models.zero_order import NanofiltrationZO
+from watertap.core.wt_database import Database
+from watertap.core.zero_order_properties import WaterParameterBlock
+
+solver = get_solver()
+
+
+class TestNFZO:
+    @pytest.fixture(scope="class")
+    def model(self):
+        m = ConcreteModel()
+        m.db = Database()
+
+        m.fs = FlowsheetBlock(default={"dynamic": False})
+        m.fs.params = WaterParameterBlock(
+            default={"solute_list": ["sulfur", "toc", "tss"]})
+
+        m.fs.unit = NanofiltrationZO(default={
+            "property_package": m.fs.params,
+            "database": m.db})
+
+        m.fs.unit.inlet.flow_vol.fix(10)
+        m.fs.unit.inlet.conc_mass_comp[0, "sulfur"].fix(1)
+        m.fs.unit.inlet.conc_mass_comp[0, "toc"].fix(2)
+        m.fs.unit.inlet.conc_mass_comp[0, "tss"].fix(3)
+        m.fs.unit.inlet.pressure.fix(1e5)
```

Comment:
There's a problem here: we're fixing volumetric flow rate in (pump flow) and pressure (pump pressure), and then we fix "electricity_intensity" after loading parameters from the database. "electricity_intensity" in this case should be equal to flow_vol * pressure / pump_efficiency / flow_vol multiplied by a conversion factor to convert the whole thing to kWh/m3. I see some issues:

(1) The fixed value for electricity_intensity doesn't equal the calculation mentioned above. The NF unit model should have a constraint linking electricity_intensity to flow_vol in and pressure in. 
(2) Pump efficiency should probably be added to the unit model as a parameter, unless we're just going to assume 100% pump efficiency all the time and don't want to allow flexibility to change that.
(3) Later on, when we want to be able to swap WT3 and Plib models, we'd have to be careful since WT3 models, more specifically, this NF model based on the WT3 approach, rolls the energy required for pumping into the unit model. Meanwhile, if we were using Plib's NF model, we'd add a pump unit model to the flowsheet before the NF model and compute the energy requirement for the pump separately. Thus, say the user wanted to swap out the WT3 version of NF from a flowsheet containing several unit models and replace it with a Plib NF model. They'd have to know/remember that energy consumption is computed differently between the two and would not be able to just swap NF models; they'd need to add the corresponding feed pump with the NF model. Looking at it the other way around, if they wanted to replace a Plib NF model on a flowsheet (which would also include the pump unit model upstream of that NF, if setup correctly), swapping the Plib NF for a WT3 NF without removing the pump from the flowsheet would lead to double-counting the energy requirement. This opens up more room for error, so we'll have to keep this in mind moving forward and decide on how to remedy this, whether through good documentation or by altering the way in which we are implementing these models.

Reply:
Remember that each unit can decide what parameters are important to it, and thus that the `yml` files do not all need to look the same. - that means that for pumps we would likely not have an energy intensity parameter, but rather an efficiency parameter instead. Additionally, a pump is a single inlet-single outlet unit, so it does not make sense to use the SITO base class for it or to define recovery and removal fractions.

Reply:
That makes sense, but we still need to address points 1 - 2 mentioned above. Furthermore, I still think we'll need to be careful when it comes time to swap these SITO unit models with Proteus unit models for the reasons mentioned. Right now, it looks like the user would exclude an IDAES pump (pressure exchanger) from the flowsheet and only include the SITO nanofiltration model; if they wanted to transition to a more detailed NF model, they'd need to make sure to add the IDAES pump unit explicitly in addition to replacing the SITO NF unit with the more detailed NF unit.

Reply:
I am not sure we need that, and going down that route might add more complexity than we want. For the NF case (and others like it), "energy intensity" is essentially an aggregate term for the efficiency of the pump(s) and the defined pressure drop across the unit - so points 1 and 2 are implicitly captured there already. Note that as energy intensity has units of power/flow, the flow effects are already accounted for.

The alternative would be to remove energy intensity entirely and instead require a pump efficiency and a pressure drop be defined for these types of units, and to then do a proper pump calculation for each. That is not hard to do, but it is starting to move us away from the original "zero order" concept where there were just aggregate parameters for everything.

Reply:
Additionally, this would mean that different types of units would need different energy calculations, adding more differences in model structure. E.g. in a stirred tank the energy requirement is not related to the pressure drop.

Reply:
Right- the energy intensity would just be pump pressure * pump flow / pump efficiency/ pump flow (assuming pump efficiency encapsulates the overall efficiency of motor and pump). However, when you are fixing energy intensity to some value, it needs to match up with the applied pressure, which isn't the case in the test shown here. So I think energy intensity should not be a loaded parameter since we will be fixing pressure to some value that may not match up with that loaded value. The nanofiltration model should probably have a constraint/expression for energy intensity.

Reply:
Another approach if we want to keep energy intensity in the yaml file-- add the default feed pressure associated with the energy intensity to the yaml file, which would then be fixed to the inlet pressure.

Reply:
There is a catch there - the pump is included within the unit for these cases, so the inlet pressure is not equal to the pump pressure (otherwise the pump would have to be external to the unit). In fact, that raises question about exactly what "pressure drop" means in these types of models as most of the pressure changes occur within the bounds of the unit.

Reply:
However, yes, pressure drop (in terms of the difference between inlet and outlet state) needs to be part of the yml file.

Reply:
@adam-a-a As you said, WT3 models generally did not separate out pumps, so their nanofiltration model includes the pump. We should keep it at that level. It would be more confusing if the user had to add a pump before nanofiltration, but they have no idea what the pressure should be fixed to (this is a zero order model so that should be specified, see paragraph below) and since the pump isn't a part of the nanofiltration model we don't have a path to specify that pump as a nanofiltration pump. I don't think we want to separate out the pumps just so that users know if they change models they need to add an IDAES pump, especially if the user isn't supposed to change how the pump operates.

I think WT3 NF model has to be for a specific pressure, otherwise it would be important to include how that changes the flux/membrane area/capital cost. So the pressure shouldn't be a degree of freedom for the user or the model and should just be unlinked to the energy consumption. If the nanofiltration model was standalone (no other units), then the pressures should be stale variables (i.e. uninvolved in any constraints). If the nanofiltration model is connected to other models, the inlet pressure should probably be fixed to atmospheric (otherwise the energy consumption data would be off), the outlet permeate pressure should probably be fixed to atmospheric (since that should be the default for filtration processes), the outlet brine pressure should be fixed to whatever value is specific to the data if we have it (it's likely WT3 doesn't have the information, so maybe atmospheric is the best value and our documentation would just need to warn users that the outlet pressure is not accurate for downstream analysis)

Reply:
The inlet pressure of the membrane module is equal to the pump discharge pressure (which is what I meant by "pump pressure", but really the "pump pressure" in the energy calculation is the added pump head, i.e., pump discharge head - pump inlet head). The pump inlet pressure is typically atmospheric while the discharge pressure is whatever pressure is needed to overcome osmotic pressure at all points along the membrane module.

The pressure drop in case of NF is the drop across the membrane module, which is only important for ensuring that the applied pressure at the module outlet (after the pressure drop) is still high enough to overcome osmotic pressure at the outlet to maintain permeate flux and ensure that there is no flux reversal.